### ISSUE TITLE
[RFC] Defining styles inline

### ISSUE THREAD

**Issue by nmn** [facebook/stylex#534]
### Describe the feature request

# Motivation

It has often been suggested that it should be possible to define styles inline. There are many reasons why such a feature might be desirable:
1. Defining styles with `stylex.create` is too much ceremony for one-off styles
2. `stylex.create` doesn't benefit for Prop type constrains when passing styles as props.
3. Developer preference for *more* co-location.

# Naming Bikeshed

There has been much discussion about what the function name for this API should be.
1. 🤏 `stylex.inline` - Slightly verbose, slightly unclear.
2. ⚛ `stylex.atom` - Terse but unclear
3. 1️⃣ `stylex.createOne` - Very verbose but clear
4. 🧑‍🍳`stylex.make` - Short and clear, but confusing alongside `stylex.create`.

I would love feedback to help make a decision here.

# Proposed API

Assuming the `stylex.inline` name, the usage would still be what you would expect from a hypothetical `stylex.createOne` function. i.e. Instead of defining:

```tsx
const styles = stylex.create({
  base: {color: 'red'}
});
```

You'd be able to do:

```tsx
const red = stylex.inline({color: 'red'})
```

Of course, it would also be possible to use this function call *directly*
with a `stylex.props()` call or a component prop.

```tsx
<div {...stylex.props(
  stylex.inline({color: 'red'})
)} />;

<MyButton styles={stylex.inline({color: 'red'})} />;
```

Since the function accepts an object, it would be possible to define a whole object of styles.

```tsx
stylex.inline({
  color: 'red',
  backgroundColor: 'black',
  textDecoration: {
    default: 'none',
    ':hover': 'underline',
    ':focus-visible': 'underline',
  },
  ...
})
```

# Constrains and Considerations

One of the biggest concerns when implementing this feature is dynamic styles. `stylex.create` enforces the usage of functions to define styles that can be dynamic. However, doing the same with the new API would be awkward.

```tsx
<MyButton styles={stylex.inline((color) => ({color}))(dynamicColor)} />;
```

Instead, the developer expectation would be the ability to use dynamic values directly within the objects

```tsx
<MyButton styles={stylex.inline({color: dynamicColor})} />;
```

This has two issues:

**One**: It is difficult to implement this correctly and detect all the ways the value of the styles may be known statically. We don't accidentally want to generate dynamic styles when the styles are actually statically know.

**Two**: It makes it far too easy to define dynamic styles. One of the reasons for requiring functions within `stylex.create` to define dynamic styles is that the vast majority of use-cases should not depend on dynamic styles. Instead conditional styles, and compositions should be sufficient. Requiring functions makes the need for truly dynamic styles explicit and enforces intentionality. Automatically generating dynamic styles with the new API would encourage patterns that would lead to bloated and slower styles.

Therefore, the current proposal would *disallow the usage of dynamic styles* within the new API entirely. Using dynamic values of styles within a `stylex.inline()` (or whatever we call it) call would be a compile-time error. It would still be possible to define constants and use them, but the amount expressivity possible would be limited by the capabilities of the compiler.

# Thoughts?

I'm posting this proposal early to gather feedback and help improve this idea before implementation.

**Comment by nmn** [comment]
`stylex.inline`

**Comment by nmn** [comment]
`stylex.atom`

**Comment by nmn** [comment]
`stylex.createOne`

**Comment by nmn** [comment]
`stylex.make`

**Comment by zaydek** [comment]
This is exciting. My only hesitation with `inline` is that a user could lead users to unintentionally believe that pseudos are not supported. Something like `atom` makes sense to me for this reason. Then you can think of StyleX as a composition of Style atoms. Either way, excited for this to land.

**Comment by necolas** [comment]
The proposed API would add significant maintenance complexity and needless surface area to StyleX. We have a design principle in React which is to only have one way to do things, even if it is more verbose in some cases. Defining styles directly on an element only looks readable in the most simple of cases. We already have 3 functions to resolve styles (one is legacy) and 2 ways to merge them (include and arrays). We could be simplifying the API rather than adding more ways to do the same thing

**Comment by zaydek** [comment]
> The proposed API would add significant maintenance complexity and needless surface area to StyleX. We have a design principle in React which is to only have one way to do things, even if it is more verbose in some cases. Defining styles directly on an element only looks readable in the most simple of cases. We already have 3 functions to resolve styles (one is legacy) and 2 ways to merge them (include and arrays). We could be simplifying the API rather than adding more ways to do the same thing

If flexibility is paramount would it not make sense to implement your concerns as ESLint warnings? The reality of not implementing `inline` means that users will defer to using inline styles rather than something that StyleX controls. I'd argue between those two choices, support for `inline` is probably preferable over users implementing inline styles themselves. Of course, users will probably still implement inline styles (unintentionally), but at least this way there's a small, incremental path to refactoring their code which is still correct.

Another implementation would be having strict mode, which disciplines developers to write code one way, like @necolas is suggesting. But not having a need for strict mode is even better.

For what it's worth I'd be willing to give up `stylex.include` for support for `inline`. I found I can lean into some kind of pattern like this (particularly with typography), when I want to compose styles:

```tsx
import * as stylex from "@stylexjs/stylex";

export type ExtendProps = { extend?: stylex.StyleXStyles };

export function Title({ extend, ...props }: ExtendProps & JSX.IntrinsicElements["div"]) {
  return <div {...props} {...stylex.props(styles.title, extend)} />;
}
```

Then I can simply use `<Title extend={styles.foo} />`. I imagine RSD has some way of making this simpler, still.

Anyway, if StyleX makes `inline` or something like it a first-class primitive, I can imagine it being easily abused. A lot of people would love to implement styles orthogonal to the way they write Emotion CSS and Tailwind, which is usually more productive in the short term and easily becomes a clusterfuck of unreadable code down the road.

For this reason, the StyleSheet API, although somewhat ceremonious at times, reinforces the idea that correctness is about narrowing decisions, which I certainly appreciate. A little bit of pain up front can be worth if it if it narrows the overall complexity space of what you're trying to achieve.

@nmn Does `inline` also work in React Native? Another question would be, would this be possible to implement as a standalone package experiment?

**Comment by nmn** [comment]
> The proposed API would add significant maintenance complexity and needless surface area to StyleX.

I share this concern. I have received repeated requests for something on these lines and so I'm trying to gather feedback to see if makes sense. The "Constrains & Considerations" section also describes some limitations that make me way of implementing this API.

----

@zaydek This **RFC** should not yet be considered as something we're about to land. This issue exists to gather feedback. There are still many reasons against doing this. The benefits would need to great outweigh those downsides.

>  if StyleX makes inline or something like it a first-class primitive, I can imagine it being easily abused.

This is another valid reason against adding this API. Although a bit subjective, I believe enforcing the usage of `stylex.create` for all styles results in more readable styles in the long run.

> would this be possible to implement as a standalone package experiment?

This is possible and what I'm working towards. I don't want to bring this into "core" anytime soon.

**Comment by zaydek** [comment]
> Did you mean RSD? RN supports inline style objects. RSD does not support this API.

I actually meant RN but I didn't know RN supports inline style objects already. Also, it sounds like RSD is effectively strict mode for StyleX so that seems aligned.

Thanks for clarifying this is not 'about to land' as I assumed. FWIW everytime I've tried to write inline style-based UIs first it's always come back to haunt me. It tooks some getting used to StyleX's StyleSheet approach but honestly it's really just like CSS Modules but typed and colocated. Now it's a no brainer for me to use StyleX and I love the API. So a little friction is fine and totally valid if you're trying to do something different, which StyleX is.

Experiments help gather information, so perhaps there's some opportunity that `inline` _unlocks_, but given the choice, I think I'm indifferent. I already learned my lesson with `stylex.include` that I don't really need it, even though originally I thought I did. I suspect the same could be true for `inline`.

I do use inline styles alongside StyleX at present, but it's simply because I'm too lazy. It's probably only 2-5% so it doesn't really matter either way. I genuinely appreciate there are so few ways to do things in StyleX, as [this tweet reads](https://twitter.com/stylexjs/status/1758396123910390110):

> The only concepts you need to understand in StyleX:
>
> 1. Style Objects - apply them, merge them, use them conditionally whatever
> 2. Variable Objects - use them within styles
> 3. Themes - set values for variables on a UI sub-tree
>
> That’s it. Seriously.

This is very good 'north star' for StyleX.

**Comment by nonzzz** [comment]
I have been using stylex for a long time,To be honest, the current method is too cumbersome for react, So i mad a babel-plugin for this problem. [test-case](https://github.com/nonzzz/stylex-extend/blob/main/packages/babel-plugin/__tests__/fixtures/stylex-macro/dynamic-spread-prop/code.js)

You might notice i'm using `JSXAttributes`, In my opinion this is simpler to use.

**Comment by nonzzz** [comment]
AFAIK. The dynamic values are very difficult to calculate, So i have to treat each attribute as a token. But i don't think is a good way. I try to merge static token to a whole object until the token meet dynamic or `SpreadElement` node. But I don't know if it's worth it.

**Comment by nmn** [comment]
> Why can't you put style objects directly in the stylex.props call?
@aspizu This isn't feasible for a couple of reasons:
1. It only works for styles being applied directly on an element and not props to custom components
2. Since conditions can be used within `stylex.props` detecting inline styles without a function call wrapper can be error prone.

**Comment by necolas** [comment]
> The reality of not implementing inline means that users will defer to using inline styles rather than something that StyleX controls

Is the proposed name (`inline`) is confusing you (or me)? This isn't an RFC for adding what is traditionally referred to as "inline styles", we already support those using the function syntax within `create()`.

> Then I can simply use <Title extend={styles.foo} />. I imagine RSD has some way of making this simpler, still.

No need for a new prop name, we just pass styles down using the `style` prop.

**Comment by BMCwebdev** [comment]
Hey @nmn, following up from my discussion post. If this veers too far from this RFC topic, I can move it back to a discussion.

I appreciate the current `stylex.create` method. It works for my use case: maintainer of a Design System and Component Library that is only used internally. I’m not making something to release to the world. 

In particular I am talking about when a Stylex object is created by a consumer of my library and passed back to the component. For example, I may have a component like this
```
export interface AcmeBoxProps {
  userStyles?: StyleXStyles<{
    backgroundColor: 'red' | 'blue' | 'green' | 'yellow' | 'purple';
  }>
}

export function AcmeBox({userStyles}: AcmeBoxProps) {
  return (
    <div
      {...stylex.props(
        userStyles,
      )}
    >
     ...
    </div>
  );
}
```

And they may use it like this
```
const styles = stylex.create({
  base: {
    backgroundColor: 'yellow',
  }
});

function App() {
  return (
    <AcmeBox userStyles={styles.base} />
  );
}
```

I may set up one of my component to accept a Stylex style object for several reasons
* I’m trying to avoid dynamic styles for obvious reasons but also because I don’t know if my company, which is currently ramping up their Content Security Policies, will decide to disallow inline styles.
* The component may be something like a Columns layout component, and the value for, say `width`, needs to accept anything. It can’t be constrained to pre-defined choices.
* Sticking with the Columns component, perhaps I am using something like CSS Grid, and I need to allow the user to pass in a large variety of values, numbers/strings for names, column definitions, etc.
* Having mentioned the need for elasticity, I also *love* that you can constrain the style object to only accept certain properties, and then further only certain values for those properties. 

Where I run into an issue in this flow is documentation, demonstrations, and developer experience. Conditional and Variant styles work great. Because they are just props, with typed options, this will all get picked up by Storybook. A consumer of my library can refer to the Storybook to see all the options available to them. They can also select one of the typed values from an auto-generated dropdown to see the resulting UI. While writing their app code, and using my component, they will get auto-complete and suggestions in their IDE for variant names and typed value options.

However, if I am expecting the user to pass in a Stylex style object, this won’t get picked up by Storybook automatically. There will be no developer experience of auto-complete or suggestions in the IDE. Even if I constrain my Stylex CSS properties and values (as in the first code snippet above).

The user will have to know that a style object is expected. They have to know what properties are allowed. They have to know what values are allowed. They will have to look back and fourth between wherever this is documented and their own code. And if that place is something like a Storybook, there should also be a way for them to manipulate the options to see the resulting UI.

This is where I am at right now. I’m moving forward with trying to only use Conditional  and Variant styles. But I know I’ll hit this wall at some point and will need to figure out a solution. I am very curious as to what people working at Meta have been doing. I assume you roll your own “Storybook” documentation solution. Even still, it’s a bummer that even while we are using TypeScript for properties and values constraints, the IDE feedback doesn’t occur in these style objects.

**Comment by nmn** [comment]
@BMCwebdev This is a broader discussion that should live in "discussions". This issue is for discussing the possibility of adding an API for defining styles "inline" only.

**Comment by tounsoo** [comment]
I like the name `inline` but I agree that this might not be a good feature to take on at the moment - maybe in the future? I share the same concerns and also like having only one way of doing things as much as possible.

**Comment by stevewillcock** [comment]
This is interesting. I'm not sure that it's needed but I can see it would add a bit of convenience.

As for names, what about `stylex.one` - that seems both terse and clear (to me at least!) :)

**Comment by zaydek** [comment]
> > The reality of not implementing inline means that users will defer to using inline styles rather than something that StyleX controls
> 
> Is the proposed name (`inline`) is confusing you (or me)? This isn't an RFC for adding what is traditionally referred to as "inline styles", we already support those using the function syntax within `create()`.
> 
> > Then I can simply use <Title extend={styles.foo} />. I imagine RSD has some way of making this simpler, still.
> 
> No need for a new prop name, we just pass styles down using the `style` prop.

The name is only confusing to me because `stylex.inline` doesn't necessarily communicate that `stylex.inline` can do _more_ than inline styles traditionally can (pseudos). But this is a minor point.

I think the question to ask is, _is something like this:_

```tsx
<div {...stylex.props(styles.foo, stylex.inline({ color: red }))} />
```

_worth not having the confidence that all styles are implementing using `stylex.create`?_

I admit you changed my mind that having multiple ways to do the same thing is probably not worth it. You could end up with code bases that are 90% `stylex.create` or conversely 90% `stylex.inline`, which I think could be worse than just not supporting `stylex.inline` at all.

I personally really appreciate the idea that there's only one correct way to do things. Even if it's more painful to learn the API initially, it's worth it in the long run to constrain unnecessary complexity.

**Comment by prc5** [comment]
>Leads to less readable code - Enforcing all styles be defined in stylex.create arguably creates more readable and maintainable code in the long run at the cost of some minor inconvenience upfront.

I think this can be enforced by linter setup and decided by end users if they want to block it or not. 

Maybe also some size checking if it exceeds x rows/styles then it should throw an error `Error: Use stylex.create for larger style collections`?

**Comment by nmn** [comment]
@prc5 Using `cva` with StyleX is an anti-pattern as the extra complexity isn't needed. We've covered alternatives that fit with StyleX better before.

**Comment by CanRau** [comment]
Here's some reference in the docs https://stylexjs.com/docs/learn/styling-ui/using-styles/#style-variants @prc5 think that's what Naman is referring to

**Comment by prc5** [comment]
@CanRau Thank you reference. I used cva here as a bit of a shortcut to show that it is possible to use the given logic in some way. I myself think that since Stylex itself allows you to build various types of abstractions around it - after all, it is JS that we know well - maybe it's just fine to allow it?

@nmn My observations are that with larger components, some logic overhead may be useful. After some time, initializing everything manually and assigning becomes just tiring, there are a lot of disconnected variables around - hence my experiments with some logic to organize it (and force it). I guess, I still haven't found the sweet spot for organizing these styles. I understand that this approach may cause unpredictable side-effects, hence the name "anti-pattern". 

I see `stylex.inline` at the same time as a method to easily abuse the way of using Stylex and the method that can increase the ease of use.

Nevertheless, I hope that this will bring you at least a little closer to the final decision. Good luck!

**Comment by nmn** [comment]
> after all, it is JS that we know well - maybe it's just fine to allow it?

1. It's important to remember that StyleX is a compiler and does not allow arbitrary Javascript. All patterns need to static.
2. Secondarily, our core principle is to keep the API simple and provide one flexible API rather than every possible API that someone might want. We're very happy with the variants patterns that can be achieved with a simple object and don't want to introduce the non-standard "cva" pattern just because it's currently a popular pattern in other styling solutions. (Objects are forever!)

> I guess, I still haven't found the sweet spot for organizing these styles

Please open a discussion topic with your questions. In our experience using StyleX at Meta for about 5 years now, we find that while sometimes StyleX can be a bit more verbose up-front, it can be used to create readable, maintainable styling with very complex logic.

The requirement to name styles also helps readability as you can focus on "what" rather than "how" when reading the styles of a component and focus on the logic to apply styles conditionally.

**Comment by zaydek** [comment]
I just spent the last hour detailing why I think StyleX should implement `stylex.inline` or `stylex.atom` and I somehow accidentally deleted my clipboard buffer. Oh well. I'll try to summarize my thoughts here.

At present, when implementing StyleX, you are expected to use the stylesheet API, also known as `stylex.create`. This is good because there are few ways to do things, but it becomes very problematic because it _requires_ you to step back and think about the taxonomy of your application as opposed to just building it. It's pretty obvious that libraries like Tailwind are popular because they don't force you to think, but they do lead to pretty obfuscated code that gets gnarlier as the project grows. I did an experiment last night where I attempted to combine the benefits of Tailwind and StyleX and I'm really happy with the results.

In short, you can express utilities in StyleX by enumerating all the properties and values per property that you care about. 

For example:

```tsx
export const display = stylex.create({
  flex: { display: "flex" },
  grid: { display: "grid" },
  arbitrary: (value: CSSProperties["display"]) => ({ display: value }),
});

export const flexDirection = stylex.create({
  row: { flexDirection: "row" },
  column: { flexDirection: "column" },
  rowReverse: { flexDirection: "row-reverse" },
  columnReverse: { flexDirection: "column-reverse" },
  arbitrary: (value: CSSProperties["flexDirection"]) => ({ flexDirection: value }),
});

export const justifyContent = stylex.create({
  normal: { justifyContent: "normal" },
  start: { justifyContent: "flex-start" },
  end: { justifyContent: "flex-end" },
  center: { justifyContent: "center" },
  spaceBetween: { justifyContent: "space-between" },
  spaceAround: { justifyContent: "space-around" },
  spaceEvenly: { justifyContent: "space-evenly" },
  stretch: { justifyContent: "stretch" },
  arbitrary: (value: CSSProperties["justifyContent"]) => ({ justifyContent: value }),
});

export const alignItems = stylex.create({
  start: { alignItems: "flex-start" },
  end: { alignItems: "flex-end" },
  center: { alignItems: "center" },
  baseline: { alignItems: "baseline" },
  stretch: { alignItems: "stretch" },
  arbitrary: (value: CSSProperties["alignItems"]) => ({ alignItems: value }),
});

export const gap = stylex.create({
  0: { gap: 0 },
  4: { gap: 4 },
  8: { gap: 8 },
  12: { gap: 12 },
  16: { gap: 16 },
  20: { gap: 20 },
  24: { gap: 24 },
  32: { gap: 32 },
  40: { gap: 40 },
  48: { gap: 48 },
  56: { gap: 56 },
  64: { gap: 64 },
  arbitrary: (value: CSSProperties["gap"]) => ({ gap: value }),
});
```

And then you export an object that wraps all of these utilities:

```tsx
export const util = {
  display,
  flexDirection,
  justifyContent,
  alignItems,
  gap,
  ...
}
```

Then you can very easily express what would otherwise be small stylesheet objects as a composition of a few utilities.

Here is some code I wrote last night:

```tsx
export function CheckboxItem() {
  const [checked, setChecked] = useState<Checked>("UNCHECKED");

  return (
    <button
      {...stylex.props(styles.touchableContainer, util.alignItems.center, util.display.flex, util.gap[8])}
      onClick={() => setChecked(toggleCheckedStart)}
    >
      <div
        {...stylex.props(
          util.backgroundColor.arbitrary(varOf(touchableBackgroundColor)),
          util.borderRadius[1e3],
          util.display.flex,
          util.flex[1],
          util.minInlineSize[0],
          util.paddingInlineEnd[16],
        )}
      >
        <div {...stylex.props(util.padding[8])}>
          <div
            {...stylex.props(
              util.backgroundColor.arbitrary("rgba(0 0 0 / 0.5)"),
              util.blockSize[16],
              util.borderRadius[1e3],
              util.inlineSize[16],
            )}
          />
        </div>
        <div {...stylex.props(util.alignItems.center, util.display.flex, util.flex[1], util.minInlineSize[0])}>
          <div {...stylex.props(util.overflow.hidden, util.textOverflow.ellipsis, util.whiteSpace.nowrap)}>
            Hello Hello Hello Hello Hello Hello Hello Hello Hello Hello Hello Hello
          </div>
        </div>
      </div>
      <Checkbox checked={checked} setChecked={setChecked} />
    </button>
  );
}
```

This is kind of mind-blowing to me because it's like [typewind](https://typewind.vercel.app/) but _is_ StyleX. Meaning that, property names are the same as the ones you'd be expected to author using the stylesheet API, you can artificially constrain values, you can still opt into arbitrary values, and _last still wins_. Though I haven't tested it, I suspect this will 'just work' in React Native, too. So arguably there is less learning involved when you consume this type of API, it's very _unsurprising_. There's no shorthands or aliases. You can just keep writing code that is orthogonal to the way the stylesheet API works. And when you need to express an arbitrary value that can't be enumerated, you just use `<property>.arbitrary(value)` and StyleX does the trick behind the scenes to resolve this as a CSS variable so it all works out. What I also love about this is that formatting tools like Prettier and dprint will automatically format long lines of code as vertical lists which I think is that much easier to read and maintain, and it's very easy to quickly alphabetically sort properties (using Sort Lines Ascending in VSCode).

Of course, this is slightly terse but it doesn't rely on extensions and you can artificially constrain allowed values, like creating a numerical system for incrementing by `4`, then `8`, then `16` and so on, meaning that you only need to expose as much complexity as you want to implement. This is the one advantage I see something like this has _over_ `stylex.inline` or `stylex.atom` because it means you can limit the range of possible values. But that's really just an aside.

I really enjoy this because it means that I can incrementally adopt the stylesheet API only when I truly need to compose 5+ properties or media queries or variants in one place, but most of the time, maybe as much as half the time, I don't want to go through all the unnecessary ceremony of creating a stylesheet object and naming things. It's just too much work when what you're trying to express isn't actually intrinsically complicated.

So my encouragement is the following, a `stylex.inline` or `stylex.atom` API would make the following possible:

* Allow users to incrementally adopt the stylesheet API when needed
* Allow users to decompose simple style objects as a composition of utilities
* Offer feature parity with CSS in JS libraries that expose `className={css(...)}` as a primitive
* Offer the benefits of utility styles but with the safety and sanity of TypeScript and no extensions necessary
* Offer the ability to inline media queries and breakpoints (which my naive implementation can't do)

As far as the original argument, which was that StyleX _should not_ adopt inline styles because it would mean there are multiple ways to do the same thing, I actually think there is a nuance not being said. That is, with the stylesheet API, _there are still multiple ways to name things_. Naming things may be _more_ useful in a team but when you're working solo or in a small team I don't think the benefits are obvious. Naming things is an exercise in taxonomy, where you have to stop and think about what to name it and why and what story you're telling, but none of that is material when most of your styles are a simple composition of several properties. And when you do have some piece of UI that is intrinsically complicated, then you should absolutely extract code into the stylesheet API and name styles and animations and compose them because what you're trying to do is intrinsically complicated. The benefit of naming things seems to be proportional to the complexity at hand.

Right now, my experience with StyleX is that it makes 'hard things easy and easy things hard'. Conversely, Tailwind makes 'easy things easy and hard things hard'. With StyleX, I can express complicated UI in a very predictable, straightforward way. It's amazing. But when it comes to the rest of the UI, which tends to be fewer than 5 properties per element, I find the stylesheet API to be really frustrating to have to stop, step back, and think about _'What story am I telling?'_. In Tailwind, I can express most UI in fewer than 5 classes, but whenever I approach anything that is intrinsically complex it spirals out of control and I give up sanity and safety along the way. This is what led me to the observation that having something like utilities in StyleX, akin to `stylex.inline` or `stylex.atom` or my naive `util` implementation, provides me safety and sanity from the easiest to most complex UI because then I get to incrementally adopt the stylesheet API _when I need it_.

In summary, I'd love to experiment with a first-class primitive for `stylex.inline` or `stylex.atom` but I can also see myself living without it because of my naive `util` implementation. Though, I suspect we wouldn't be having this conversation if there wasn't a real need for it. I don't mind naming some things but for the life of me I really don't want to have to name everything.

**Comment by javascripter** [comment]
> Defining styles directly on an element only looks readable in the most simple of cases.

It's true that named styles can be more readable for larger components. However, we have the option to split them into separate, smaller components and name them appropriately to make them more readable as well.

Named styles can improve readability for large components:
```jsx
const styles = stylex.create({
  container: { display: 'flex', flexDirection: 'column', padding: '1rem' },
  title: { fontSize: '2rem', marginBottom: '1rem' },
  text: { fontSize: '1rem', lineHeight: '1.5' },
  button: { backgroundColor: 'blue', color: 'white', padding: '0.5rem 1rem' },
})

function LargeComponent() {
  return (
    <html.div style={styles.container}>
      <html.h1 style={styles.title}>Title</html.h1>
      <html.p style={styles.text}>Some text content...</html.p>
      <html.button style={styles.button}>Click me</html.button>
    </html.div>
  )
}
```

Splitting the component into smaller ones can also improve readability:
```jsx
function Title() {
  return <html.h1 style={stylex.inline({fontSize: '2rem', marginBottom: '1rem'})}>Title</html.h1>
}

function Text() {
  return <p style={stylex.inline({fontSize: '1rem', lineHeight: '1.5'})}>Some text content...</html.p>
}

function Button() {
  return <html.button style={stylex.inline({backgroundColor: 'blue', color: 'white', padding: '0.5rem 1rem'})}>Click me</html.button>;
}

function LargeComponent() {
  return (
    <html.div style={stylex.inline({display: 'flex', flexDirection: 'column', padding: '1rem'})}>
      <Title />
      <Text />
      <Button />
    </html.div>
  );
}
```

In practice, I use inline styles extensively in my codebase and find 3-4 levels of nesting and about 100 lines of code for each component (including inline styles) still quite manageable and readable. When things get too large or complex,splitting components often make more sense anyway because it makes it clear what parts depend on which props/state, makes it easier to reduce re-renders by memoization, etc.

In conclusion, although I understand the concern about potential increase in complexity, I find inline styles very useful. Not all styles need naming if they are short and I find style colocation more readable in many cases.

**Comment by nmn** [comment]
@zaydek This may be of interest to you: https://gist.github.com/nmn/b98c21fbf3ee02c35319940b14030f62

**Comment by nonzzz** [comment]
In fact, I tend using `inline` to define simple styles in my component like:

```tsx

import React from 'react'
import { useScale, withScale } from '../../composables'

interface Props {
  inline?: boolean
}

export type SpacerProps = Omit<React.HTMLAttributes<any>, keyof Props> & Props

function SpacerComponent({ inline = false, ...props }: SpacerProps) {
  const { SCALES } = useScale()
  return (
    <div
      stylex={{
        width: SCALES.width(1),
        height: SCALES.height(1),
        padding: `${SCALES.pt(0)} ${SCALES.pr(0)} ${SCALES.pb(0)} ${SCALES.pl(0)}`,
        margin: `${SCALES.mt(0)} ${SCALES.mr(0)} ${SCALES.mb(0)} ${SCALES.ml(0)}`,
        display: 'block',
        ...(inline && { display: 'inline-block' })
       
      }}
      {...props}
    />
  )
}

SpacerComponent.displayName = 'Spacer'

export const Spacer = withScale(SpacerComponent)


```

`JSXAtrribute` is same as inline function, In my option only simple and short style sheet need it. If us style sheet is big, and we need compose some short style sheet i prefer using 

```tsx

import * as stylex from '@stylexjs/stylex'

const styles = stylex.create({
    // ... a long style sheet
})


function Component() {

return <div {...stylex.props(styles.icon,inline({position:'static'}))}></div>

}

```

**Comment by nonzzz** [comment]
Let me explain what i did. The `JSXAttribute` is just translating `JSXAttribute` into `props` or `attrs` and combing it with `create` In the follow code block

```tsx
// input

const Component = () => {
  return <div stylex={{ color: "red" }} />;
};

// output

import { create, props } from "@stylexjs/stylex";

const styles = create({
  xxHash: {
    color: "red",
  },
});

const Component = () => {
  return <div {...props(styles.xxHash)} />;
};
```

Back to topic. In my option is that `inline` is equivalent to `create`. So i extra them to the toplevel and translate them. Here is a simple example

```tsx
// input

import { create, props } from "@stylexjs/stylex";

const styles = create({
  test: {
    // ... a long style sheet
  },
});

const Component = () => {
  return <div {...props(styles.test, inline({ position: "static" }))} />;
};

// output

import { create, props } from "@stylexjs/stylex";

const styles = create({
  test: {
    // ... a long style sheet
  },
});

const _styles = create({
  xxHash: {
    position: "static",
  },
});

const Component = () => {
  return <div {...props(styles.test, _styles.xxHash)} />;
};
```

**Comment by zaydek** [comment]
@nonzzz Is that a custom Vite plugin? What's doing the translation for you? That's really interesting.

**Comment by nonzzz** [comment]
@zaydek Yeah, I create an babel-plugin to do this. [plugin](https://github.com/nonzzz/stylex-extend/tree/main/packages/babel-plugin)

**Comment by zaydek** [comment]
<img src="https://i.kym-cdn.com/photos/images/newsfeed/001/569/898/26c.jpg">

**Comment by hrasoa** [comment]
> Just to follow up, @nonzzz did manage to implement `inline` as a Vite plugin (and it works!). I setup a simple repo to provide footing for anyone who wants to experiment with this. Exciting times.
> 
> https://github.com/zaydek/stylex-extend

Why ? it's not event spec'ed yet.

**Comment by zaydek** [comment]
I took some encouragement from this reply and wanted to double down on a simple approach: https://gist.github.com/zaydek/6a3abb586c6bb82405d4f4b9c4ce7ee2.

> The `util` example is natively supported by StyleX with the existing APIs, and results in less JS code in bundles.

If anyone's struggling with wanting to be able to more easily define inline styles in StyleX and doesn't want to depend on plugins, I've been using a simplified version of the `utils` pattern to great effect. You wouldn't want to use it everywhere because it creates CSS variables for all properties, but if there's one or two properties you'd rather not model a style object for, I find this to be very ergonomic.

This is simpler than my previous implementation because it doesn't hard code what values you can use and defers to the `CSSProperties` type for the specified property:

```ts
export const u = stylex.create({
  accentColor: (value: CSSProperties["accentColor"]) => ({ accentColor: value }),
  alignContent: (value: CSSProperties["alignContent"]) => ({ alignContent: value }),
  alignItems: (value: CSSProperties["alignItems"]) => ({ alignItems: value }),
  alignSelf: (value: CSSProperties["alignSelf"]) => ({ alignSelf: value }),
  alignTracks: (value: CSSProperties["alignTracks"]) => ({ alignTracks: value }),
  ...
});
```

Looks like this, you probably always want to specify `utils` last.

```tsx
<header {...stylex.props(dialogStyles.header, u.paddingBlockEnd(0))}>
  <div {...stylex.props(dialogStyles.hero, u.alignSelf("center"))} />
  <h1 {...stylex.props(typographyStyles.title, u.textAlign("center"))}>
    Are you sure you want to continue? This action is irreversible.
  </h1>
</header>
```

I found StyleX to be a lot easier having something like this in my back pocket, particularly for prototyping and scaffolding ideas before I'm ready to name them.

**Comment by nmn** [comment]
> This is simpler than my previous implementation because it doesn't hard code what values you can use and defers to the CSSProperties type for the specified property:

This is worse for performance as when you use dynamic styles with functions, you're relying on inline styles to set variables in addition to setting classNames.

**Comment by nonzzz** [comment]
A friendly ping. @nmn This RFC hasn't  been updated for a long time. FWIW, I implemented it and using for a long time. I found it very useful for simple cases.
Is there any further information or updates planned for the RFC?

**Comment by nmn** [comment]
@nonzzz The current decision is to not implement this in StyleX itself in order to keep the API small. But the issue remains open as a central place for people to be able to chime in with their opinions.

**Comment by frattaro** [comment]
Could you make a jsx parser to pull out inline styles? like
```jsx
<div stylex={{
  borderColor: "red"
}}>
  My div
</div>
```
that then pulls out the data, throws it into `stylex.create` and links w/`stylex.props`?

**Comment by nmn** [comment]
@frattaro Look at https://github.com/zaydek/stylex-extend

**Comment by necolas** [comment]
Moving this to Discussions
